File: prac415/mnsolutions/fourthought.java
import java.util.Scanner;

public class fourthought{
    public static void main(String[] args){
        Scanner inp = new Scanner(System.in);
        int numTrials = inp.nextInt();
        int[][][] lookup = new int[4][4][4];
        for(int i = 0; i < 4; i++){
            for(int j = 0; j < 4; j++){
                for(int k = 0; k < 4; k++){
                    int temp = 4;
                    switch(i){
                        case 0:
                            temp *= 4;
                            break;
                        case 1:
                            temp += 4;
                            // better in python with eval...
                    }
                }
            }
        }
        for(int i = 0; i < numTrials; i++){
            
        }
    }
}

File: prac415/mnsolutions/fourthought.py
numTrials = int(input())
arr = []

for i in range(0,4):
    for j in range(0,4):
        for k in range(0,4):
            temp = "4"
            if(i == 0):
                temp += " + 4"
            elif(i == 1):
                temp += " - 4"
            elif(i == 2):
                temp += " * 4"
            else:
                temp += " / 4"
            if(j == 0):
                temp += " + 4"
            elif(j == 1):
                temp += " - 4"
            elif(j == 2):
                temp +=" * 4"
            else:
                temp +=" / 4"
            if(k == 0):
                temp+=" + 4"
            elif(k == 1):
                temp +=" - 4"
            elif(k == 2):
                temp+=" * 4"
            else:
                temp+=" / 4"
            arr.append(temp)

for i in range(0,numTrials):
    num = int(input())
    found = False
    for j in range(0,64):
        if eval(arr[j]) == num:
            found = True
            break
    if(found):
        print(arr[j] + " = " + str(num))
    else:
        print "no solution"
